File: mergeSort.js
function mergeSort(aList) {
  if (aList.length <= 1) {
    return aList;
  }

  var mid = parseInt(aList.length / 2);
  var left = aList.slice(0, mid);
  var right = aList.slice(mid);

  return merge(mergeSort(left), mergeSort(right));
}


function merge(a, b) {
  var merged = [];

  while (a.length && b.length) {
    if (a[0] < b[0]) {
      merged.push(a.shift());
    } else {
      merged.push(b.shift());
    }
  }

  while (a.length) {
    merged.push(a.shift());
  }

  while (b.length) {
    merged.push(b.shift());
  }

  return merged;
}



function mergeV2(a, b) {
  var merged = [];
  var aIndex = 0;
  var bIndex = 0;

  while (aIndex < a.length && bIndex < b.length) {
    if (a[aIndex] < b[bIndex]) {
      merged.push(a[aIndex]);
      aIndex++;
    } else {
      merged.push(b[bIndex]);
      bIndex++;
    }
  }

  while (aIndex < a.length) {
    merged.push(a[aIndex]);
    aIndex++;
  }

  while (bIndex < b.length) {
    merged.push(b[bIndex]);
    bIndex++;
  }

  return merged;
}


console.log(merge([3, 4, 5], [0, 2, 7, 8]));
console.log(mergeSort([3, 4, 5, 0, 2, 7, 8]));

File: mergeSort.py
def mergeSort(a_list):
    if len(a_list) <= 1:
        return a_list

    mid = len(a_list) // 2
    left = a_list[:mid]
    right = a_list[mid:]

    return merge(mergeSort(left), mergeSort(right));


def merge(left, right):
    merged_list = []
    l_index = 0
    r_index = 0

    while l_index < len(left) and r_index < len(right):
        if left[l_index] < right[r_index]:
            merged_list.append(left[l_index])
            l_index += 1
        else:
            merged_list.append(right[r_index])
            r_index += 1

    while l_index < len(left):
        merged_list.append(left[l_index])
        l_index += 1
    while r_index < len(right):
        merged_list.append(right[r_index])
        r_index += 1

    return merged_list

print merge([3, 4, 5], [0, 2, 7, 8])
print mergeSort([3, 4, 5, 0, 2, 7, 8, 1])
